### PR TITLE
add Path property type

### PR DESCRIPTION
**Pull request by albertchen-sifive**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)
Adds new public `PathType` class that represents a path to a hardware instance or member. Adds `PathPropertyType`/`PathPropertyLiteral` and `PropertyType` instances for module/data/mem.

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x`, `3.6.x`, or `5.x` depending on impact, API modification or big change: `6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: core/src/main/scala/chisel3/properties/PathType.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    }
+  }
+
+  /** Construct a PathType that refers to a Data
+    */
+  def apply(data: Data): PathType = {
+    new PathType {
+      def toTarget(): IsMember = data.toAbsoluteTarget
+    }
+  }
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/properties/PathType.scala:42]: I'm going to approve this as is, but we really validate at this point if the `Data` can have a target. I'm mainly thinking about things like `PathType(3.U)` which will result an an exception thrown later when `.toAbsoluteTarget` is actually called.

File: core/src/main/scala/chisel3/properties/PathType.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+package chisel3.properties
+
+import chisel3.{Data, MemBase}
+import chisel3.experimental.BaseModule
+import firrtl.annotations.{InstanceTarget, IsMember, ModuleTarget, ReferenceTarget}
+import firrtl.ir.{PathPropertyLiteral}
+
+/** Represent a Path type for referencing a hardware instance or member in a Property[PathType]
+  */
+sealed abstract class PathType {
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/properties/PathType.scala:12]: Should this be `PathType` or perhaps just `Path`? It's `Path` in firrtl.

**Review by jackkoenig — APPROVED** [review]
Implementation LGTM, just wondering if we should be calling this `Path` instead of `PathType`.

File: core/src/main/scala/chisel3/properties/PathType.scala (hunk 2 above)
Comment (albertchen-sifive) [reply core/src/main/scala/chisel3/properties/PathType.scala]: renamed to `Path`

**Comment by sequencer** [comment]
Interesting. Maybe this is the real replacement to `TraceAPI`

**Pull request merged**